File: src/main/java/org/cubexmc/metro/manager/RailProtectionManager.java
package org.cubexmc.metro.manager;

import org.bukkit.Bukkit;
import org.bukkit.Material;
import org.bukkit.World;
import org.bukkit.block.Block;
import org.bukkit.entity.Entity;
import org.bukkit.entity.Minecart;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.EventPriority;
import org.bukkit.event.Listener;
import org.bukkit.event.block.BlockBreakEvent;
import org.bukkit.persistence.PersistentDataType;
import org.cubexmc.metro.Metro;
import org.cubexmc.metro.model.Line;
import org.cubexmc.metro.model.RoutePoint;
import org.cubexmc.metro.util.MetroConstants;
import org.cubexmc.metro.util.OwnershipUtil;

import java.util.ArrayList;
import java.util.Collections;
import java.util.EnumSet;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

/**
 * Builds an in-memory rail protection index from recorded route points.
 */
public class RailProtectionManager implements Listener {

    private static final double INTERPOLATION_STEP = 0.5;
    private static final Set<Material> RAIL_MATERIALS = EnumSet.of(
            Material.RAIL,
            Material.POWERED_RAIL,
            Material.DETECTOR_RAIL,
            Material.ACTIVATOR_RAIL
    );

    private final Metro plugin;
    private final Map<String, Set<BlockKey>> lineToBlocks = new HashMap<>();
    private final Map<BlockKey, Set<String>> blockToLines = new HashMap<>();
    private final Map<String, ProtectionIndexStats> lineToStats = new HashMap<>();

    public RailProtectionManager(Metro plugin) {
        this.plugin = plugin;
    }

    public synchronized void rebuildAll() {
        lineToBlocks.clear();
        blockToLines.clear();
        lineToStats.clear();
        for (Line line : plugin.getLineManager().getAllLines()) {
            rebuildLineLocked(line);
        }
    }

    public synchronized void rebuildLine(String lineId) {
        removeLineLocked(lineId);
        Line line = plugin.getLineManager().getLine(lineId);
        rebuildLineLocked(line);
    }

    public synchronized int getProtectedBlockCount(String lineId) {
        Set<BlockKey> blocks = lineToBlocks.get(lineId);
        return blocks == null ? 0 : blocks.size();
    }

    public synchronized ProtectionIndexStats getProtectionIndexStats(String lineId) {
        ProtectionIndexStats stats = lineToStats.get(lineId);
        return stats == null ? ProtectionIndexStats.empty() : stats;
    }

    private void rebuildLineLocked(Line line) {
        if (line == null || !line.isRailProtected()) {
            return;
        }
        ProtectionIndexBuilder builder = new ProtectionIndexBuilder(line.getWorldName());
        Set<BlockKey> blocks = collectRailBlocks(line.getRoutePoints(), builder);
        ProtectionIndexStats stats = builder.build(blocks.size());
        lineToStats.put(line.getId(), stats);
        if (blocks.isEmpty()) {
            return;
        }
        lineToBlocks.put(line.getId(), blocks);
        for (BlockKey block : blocks) {
            blockToLines.computeIfAbsent(block, key -> new HashSet<>()).add(line.getId());
        }
    }

    private void removeLineLocked(String lineId) {
        lineToStats.remove(lineId);
        Set<BlockKey> oldBlocks = lineToBlocks.remove(lineId);
        if (oldBlocks == null) {
            return;
        }
        for (BlockKey block : oldBlocks) {
            Set<String> lineIds = blockToLines.get(block);
            if (lineIds == null) {
                continue;
            }
            lineIds.remove(lineId);
            if (lineIds.isEmpty()) {
                blockToLines.remove(block);
            }
        }
    }

    private Set<BlockKey> collectRailBlocks(List<RoutePoint> routePoints, ProtectionIndexBuilder builder) {
        if (routePoints == null || routePoints.isEmpty()) {
            return Collections.emptySet();
        }
        Set<BlockKey> blocks = new HashSet<>();
        List<RoutePoint> points = new ArrayList<>(routePoints);
        for (int i = 0; i < points.size(); i++) {
            addNearestRail(blocks, points.get(i), builder);
            if (i + 1 < points.size()) {
                interpolateSegment(blocks, points.get(i), points.get(i + 1), builder);
            }
        }
        return blocks;
    }

    private void interpolateSegment(Set<BlockKey> blocks, RoutePoint from, RoutePoint to,
                                    ProtectionIndexBuilder builder) {
        if (from == null || to == null || !from.worldName().equals(to.worldName())) {
            return;
        }
        double dx = to.x() - from.x();
        double dy = to.y() - from.y();
        double dz = to.z() - from.z();
        double distance = Math.sqrt(dx * dx + dy * dy + dz * dz);
        int samples = Math.max(1, (int) Math.ceil(distance / INTERPOLATION_STEP));
        for (int i = 1; i < samples; i++) {
            double ratio = (double) i / samples;
            addNearestRail(blocks, new RoutePoint(
                    from.worldName(),
                    from.x() + dx * ratio,
                    from.y() + dy * ratio,
                    from.z() + dz * ratio
            ), builder);
        }
    }

    private void addNearestRail(Set<BlockKey> blocks, RoutePoint point, ProtectionIndexBuilder builder) {
        builder.sample(point);
        if (point == null) {
            return;
        }
        if (builder.isWorldMismatch(point)) {
            builder.skippedWorldMismatch();
            return;
        }
        World world = Bukkit.getWorld(point.worldName());
        if (world == null) {
            builder.skippedMissingWorld();
            return;
        }
        BlockKey nearestRail = findNearestRail(point);
        if (nearestRail != null) {
            blocks.add(nearestRail);
        } else {
            builder.skippedNoRail();
        }
    }

    private BlockKey findNearestRail(RoutePoint point) {
        World world = Bukkit.getWorld(point.worldName());
        if (world == null) {
            return null;
        }

        int baseX = (int) Math.floor(point.x());
        int baseY = (int) Math.floor(point.y());
        int baseZ = (int) Math.floor(point.z());
        BlockKey best = null;
        double bestDistance = Double.MAX_VALUE;

        for (int dy : new int[] {0, -1}) {
            for (int dx = -1; dx <= 1; dx++) {
                for (int dz = -1; dz <= 1; dz++) {
                    int x = baseX + dx;
                    int y = baseY + dy;
                    int z = baseZ + dz;
                    Block block = world.getBlockAt(x, y, z);
                    if (!isRail(block)) {
                        continue;
                    }
                    double distance = distanceSquaredToBlockCenter(point, x, y, z);
                    if (distance < bestDistance) {
                        bestDistance = distance;
                        best = new BlockKey(world.getName(), x, y, z);
                    }
                }
            }
        }
        return best;
    }

    private double distanceSquaredToBlockCenter(RoutePoint point, int x, int y, int z) {
        double dx = point.x() - (x + 0.5);
        double dy = point.y() - (y + 0.5);
        double dz = point.z() - (z + 0.5);
        return dx * dx + dy * dy + dz * dz;
    }

    @EventHandler(priority = EventPriority.HIGHEST, ignoreCancelled = true)
    public void onBlockBreak(BlockBreakEvent event) {
        Player player = event.getPlayer();
        Block block = event.getBlock();

        if (isMetroPassengerBreakingRail(player, block)) {
            event.setCancelled(true);
            player.sendMessage(plugin.getLanguageManager().getMessage("protection.passenger_break_denied"));
            return;
        }

        Set<String> protectedLineIds = getProtectedLines(block);
        if (protectedLineIds.isEmpty()) {
            return;
        }
        if (canBreakProtectedRail(player, protectedLineIds)) {
            return;
        }

        event.setCancelled(true);
        player.sendMessage(plugin.getLanguageManager().getMessage("protection.rail_break_denied"));
    }

    private boolean isMetroPassengerBreakingRail(Player player, Block block) {
        if (!plugin.getConfigFacade().isSafeModePassengerRailBreakProtection() || !isRail(block)) {
            return false;
        }
        Entity vehicle = player.getVehicle();
        if (!(vehicle instanceof Minecart minecart)) {
            return false;
        }
        return minecart.getPersistentDataContainer().has(MetroConstants.getMinecartKey(), PersistentDataType.BYTE);
    }

    private synchronized Set<String> getProtectedLines(Block block) {
        if (!isRail(block)) {
            return Collections.emptySet();
        }
        Set<String> lineIds = blockToLines.get(BlockKey.fromBlock(block));
        return lineIds == null ? Collections.emptySet() : new HashSet<>(lineIds);
    }

    private boolean canBreakProtectedRail(Player player, Set<String> protectedLineIds) {
        if (OwnershipUtil.hasAdminBypass(player)) {
            return true;
        }
        for (String lineId : protectedLineIds) {
            Line line = plugin.getLineManager().getLine(lineId);
            if (line != null && !OwnershipUtil.canManageLine(player, line)) {
                return false;
            }
        }
        return true;
    }

    private boolean isRail(Block block) {
        return block != null && RAIL_MATERIALS.contains(block.getType());
    }

    public record ProtectionIndexStats(int sampledPoints, int indexedBlocks, int skippedWorldMismatch,
                                       int skippedMissingWorld, int skippedNoRail) {
        public static ProtectionIndexStats empty() {
            return new ProtectionIndexStats(0, 0, 0, 0, 0);
        }

        public int skippedTotal() {
            return skippedWorldMismatch + skippedMissingWorld + skippedNoRail;
        }

        public boolean hasWarnings() {
            return skippedTotal() > 0;
        }
    }

    private static final class ProtectionIndexBuilder {
        private final String lineWorldName;
        private int sampledPoints;
        private int skippedWorldMismatch;
        private int skippedMissingWorld;
        private int skippedNoRail;

        private ProtectionIndexBuilder(String lineWorldName) {
            this.lineWorldName = lineWorldName;
        }

        private void sample(RoutePoint point) {
            if (point != null) {
                sampledPoints++;
            }
        }

        private boolean isWorldMismatch(RoutePoint point) {
            return point != null
                    && lineWorldName != null
                    && !lineWorldName.isBlank()
                    && !lineWorldName.equals(point.worldName());
        }

        private void skippedWorldMismatch() {
            skippedWorldMismatch++;
        }

        private void skippedMissingWorld() {
            skippedMissingWorld++;
        }

        private void skippedNoRail() {
            skippedNoRail++;
        }

        private ProtectionIndexStats build(int indexedBlocks) {
            return new ProtectionIndexStats(sampledPoints, indexedBlocks, skippedWorldMismatch,
                    skippedMissingWorld, skippedNoRail);
        }
    }

    private record BlockKey(String worldName, int x, int y, int z) {
        private static BlockKey fromBlock(Block block) {
            return new BlockKey(block.getWorld().getName(), block.getX(), block.getY(), block.getZ());
        }
    }
}

File: src/main/java/org/cubexmc/metro/manager/RailProtectionManager.kt
package org.cubexmc.metro.manager

import java.util.EnumSet
import org.bukkit.Bukkit
import org.bukkit.Material
import org.bukkit.World
import org.bukkit.block.Block
import org.bukkit.entity.Minecart
import org.bukkit.entity.Player
import org.bukkit.event.EventHandler
import org.bukkit.event.EventPriority
import org.bukkit.event.Listener
import org.bukkit.event.block.BlockBreakEvent
import org.bukkit.persistence.PersistentDataType
import org.cubexmc.metro.Metro
import org.cubexmc.metro.model.Line
import org.cubexmc.metro.model.RoutePoint
import org.cubexmc.metro.util.MetroConstants
import org.cubexmc.metro.util.OwnershipUtil

/**
 * Builds an in-memory rail protection index from recorded route points.
 */
class RailProtectionManager(private val plugin: Metro) : Listener {
    private val lineToBlocks: MutableMap<String, MutableSet<BlockKey>> = HashMap()
    private val blockToLines: MutableMap<BlockKey, MutableSet<String>> = HashMap()
    private val lineToStats: MutableMap<String, ProtectionIndexStats> = HashMap()

    @Synchronized
    fun rebuildAll() {
        lineToBlocks.clear()
        blockToLines.clear()
        lineToStats.clear()
        for (line in plugin.lineManager.getAllLines()) {
            rebuildLineLocked(line)
        }
    }

    @Synchronized
    fun rebuildLine(lineId: String) {
        removeLineLocked(lineId)
        rebuildLineLocked(plugin.lineManager.getLine(lineId))
    }

    @Synchronized
    fun getProtectedBlockCount(lineId: String): Int = lineToBlocks[lineId]?.size ?: 0

    @Synchronized
    fun getProtectionIndexStats(lineId: String): ProtectionIndexStats =
        lineToStats[lineId] ?: ProtectionIndexStats.empty()

    private fun rebuildLineLocked(line: Line?) {
        if (line == null || !line.isRailProtected) {
            return
        }
        val builder = ProtectionIndexBuilder(line.worldName)
        val blocks = collectRailBlocks(line.routePoints, builder)
        val stats = builder.build(blocks.size)
        lineToStats[line.id] = stats
        if (blocks.isEmpty()) {
            return
        }
        lineToBlocks[line.id] = blocks.toMutableSet()
        for (block in blocks) {
            blockToLines.computeIfAbsent(block) { HashSet() }.add(line.id)
        }
    }

    private fun removeLineLocked(lineId: String) {
        lineToStats.remove(lineId)
        val oldBlocks = lineToBlocks.remove(lineId) ?: return
        for (block in oldBlocks) {
            val lineIds = blockToLines[block] ?: continue
            lineIds.remove(lineId)
            if (lineIds.isEmpty()) {
                blockToLines.remove(block)
            }
        }
    }

    private fun collectRailBlocks(routePoints: List<RoutePoint>?, builder: ProtectionIndexBuilder): Set<BlockKey> {
        if (routePoints.isNullOrEmpty()) {
            return emptySet()
        }
        val blocks = HashSet<BlockKey>()
        val points = ArrayList(routePoints)
        for (index in points.indices) {
            addNearestRail(blocks, points[index], builder)
            if (index + 1 < points.size) {
                interpolateSegment(blocks, points[index], points[index + 1], builder)
            }
        }
        return blocks
    }

    private fun interpolateSegment(
        blocks: MutableSet<BlockKey>,
        from: RoutePoint?,
        to: RoutePoint?,
        builder: ProtectionIndexBuilder,
    ) {
        if (from == null || to == null || from.worldName() != to.worldName()) {
            return
        }
        val dx = to.x() - from.x()
        val dy = to.y() - from.y()
        val dz = to.z() - from.z()
        val distance = kotlin.math.sqrt(dx * dx + dy * dy + dz * dz)
        val samples = kotlin.math.max(1, kotlin.math.ceil(distance / INTERPOLATION_STEP).toInt())
        for (index in 1 until samples) {
            val ratio = index.toDouble() / samples
            addNearestRail(
                blocks,
                RoutePoint(
                    from.worldName(),
                    from.x() + dx * ratio,
                    from.y() + dy * ratio,
                    from.z() + dz * ratio,
                ),
                builder,
            )
        }
    }

    private fun addNearestRail(blocks: MutableSet<BlockKey>, point: RoutePoint?, builder: ProtectionIndexBuilder) {
        builder.sample(point)
        if (point == null) {
            return
        }
        if (builder.isWorldMismatch(point)) {
            builder.skippedWorldMismatch()
            return
        }
        if (Bukkit.getWorld(point.worldName()) == null) {
            builder.skippedMissingWorld()
            return
        }
        val nearestRail = findNearestRail(point)
        if (nearestRail != null) {
            blocks.add(nearestRail)
        } else {
            builder.skippedNoRail()
        }
    }

    private fun findNearestRail(point: RoutePoint): BlockKey? {
        val world: World = Bukkit.getWorld(point.worldName()) ?: return null
        val baseX = kotlin.math.floor(point.x()).toInt()
        val baseY = kotlin.math.floor(point.y()).toInt()
        val baseZ = kotlin.math.floor(point.z()).toInt()
        var best: BlockKey? = null
        var bestDistance = Double.MAX_VALUE

        for (dy in intArrayOf(0, -1)) {
            for (dx in -1..1) {
                for (dz in -1..1) {
                    val x = baseX + dx
                    val y = baseY + dy
                    val z = baseZ + dz
                    val block = world.getBlockAt(x, y, z)
                    if (!isRail(block)) {
                        continue
                    }
                    val distance = distanceSquaredToBlockCenter(point, x, y, z)
                    if (distance < bestDistance) {
                        bestDistance = distance
                        best = BlockKey(world.name, x, y, z)
                    }
                }
            }
        }
        return best
    }

    private fun distanceSquaredToBlockCenter(point: RoutePoint, x: Int, y: Int, z: Int): Double {
        val dx = point.x() - (x + 0.5)
        val dy = point.y() - (y + 0.5)
        val dz = point.z() - (z + 0.5)
        return dx * dx + dy * dy + dz * dz
    }

    @EventHandler(priority = EventPriority.HIGHEST, ignoreCancelled = true)
    fun onBlockBreak(event: BlockBreakEvent) {
        val player = event.player
        val block = event.block

        if (isMetroPassengerBreakingRail(player, block)) {
            event.isCancelled = true
            player.sendMessage(plugin.languageManager.getMessage("protection.passenger_break_denied"))
            return
        }

        val protectedLineIds = getProtectedLines(block)
        if (protectedLineIds.isEmpty() || canBreakProtectedRail(player, protectedLineIds)) {
            return
        }

        event.isCancelled = true
        player.sendMessage(plugin.languageManager.getMessage("protection.rail_break_denied"))
    }

    private fun isMetroPassengerBreakingRail(player: Player, block: Block): Boolean {
        if (!plugin.configFacade.isSafeModePassengerRailBreakProtection() || !isRail(block)) {
            return false
        }
        val minecart = player.vehicle as? Minecart ?: return false
        val minecartKey = MetroConstants.getMinecartKey() ?: return false
        return minecart.persistentDataContainer.has(minecartKey, PersistentDataType.BYTE)
    }

    @Synchronized
    private fun getProtectedLines(block: Block): Set<String> {
        if (!isRail(block)) {
            return emptySet()
        }
        val lineIds = blockToLines[BlockKey.fromBlock(block)]
        return if (lineIds == null) emptySet() else HashSet(lineIds)
    }

    private fun canBreakProtectedRail(player: Player, protectedLineIds: Set<String>): Boolean {
        if (OwnershipUtil.hasAdminBypass(player)) {
            return true
        }
        for (lineId in protectedLineIds) {
            val line = plugin.lineManager.getLine(lineId)
            if (line != null && !OwnershipUtil.canManageLine(player, line)) {
                return false
            }
        }
        return true
    }

    private fun isRail(block: Block?): Boolean = block != null && RAIL_MATERIALS.contains(block.type)

    @JvmRecord
    data class ProtectionIndexStats(
        val sampledPoints: Int,
        val indexedBlocks: Int,
        val skippedWorldMismatch: Int,
        val skippedMissingWorld: Int,
        val skippedNoRail: Int,
    ) {
        fun skippedTotal(): Int = skippedWorldMismatch + skippedMissingWorld + skippedNoRail

        fun hasWarnings(): Boolean = skippedTotal() > 0

        companion object {
            @JvmStatic
            fun empty(): ProtectionIndexStats = ProtectionIndexStats(0, 0, 0, 0, 0)
        }
    }

    private class ProtectionIndexBuilder(private val lineWorldName: String?) {
        private var sampledPoints = 0
        private var skippedWorldMismatch = 0
        private var skippedMissingWorld = 0
        private var skippedNoRail = 0

        fun sample(point: RoutePoint?) {
            if (point != null) {
                sampledPoints++
            }
        }

        fun isWorldMismatch(point: RoutePoint?): Boolean =
            point != null &&
                lineWorldName != null &&
                lineWorldName.isNotBlank() &&
                lineWorldName != point.worldName()

        fun skippedWorldMismatch() {
            skippedWorldMismatch++
        }

        fun skippedMissingWorld() {
            skippedMissingWorld++
        }

        fun skippedNoRail() {
            skippedNoRail++
        }

        fun build(indexedBlocks: Int): ProtectionIndexStats =
            ProtectionIndexStats(
                sampledPoints,
                indexedBlocks,
                skippedWorldMismatch,
                skippedMissingWorld,
                skippedNoRail,
            )
    }

    private data class BlockKey(
        val worldName: String,
        val x: Int,
        val y: Int,
        val z: Int,
    ) {
        companion object {
            fun fromBlock(block: Block): BlockKey =
                BlockKey(block.world.name, block.x, block.y, block.z)
        }
    }

    companion object {
        private const val INTERPOLATION_STEP = 0.5
        private val RAIL_MATERIALS: Set<Material> = EnumSet.of(
            Material.RAIL,
            Material.POWERED_RAIL,
            Material.DETECTOR_RAIL,
            Material.ACTIVATOR_RAIL,
        )
    }
}
